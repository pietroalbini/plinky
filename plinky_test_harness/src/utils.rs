use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
