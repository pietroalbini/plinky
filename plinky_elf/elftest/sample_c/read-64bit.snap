---
source: plinky_test_harness/src/utils.rs
---
reading ELF exited with exit status: 0

=== stdout ===
  Metadata:
╭────────────┬─────────────╮
│ Class      │ ELF 64bit   │
├────────────┼─────────────┤
│ Endian     │ Little      │
├────────────┼─────────────┤
│ ABI        │ System V    │
├────────────┼─────────────┤
│ Machine    │ x86-64      │
├────────────┼─────────────┤
│ Type       │ Relocatable │
├────────────┼─────────────┤
│ Entrypoint │ -           │
╰────────────┴─────────────╯

section #0 (address: 0x0)
 │
 │  empty section
 ┴

section .shstrtab#1 (address: 0x0)
 │
 │    Strings table:
 │  ╭──────┬─────────────────╮
 │  │ 0x0  │                 │
 │  ├──────┼─────────────────┤
 │  │ 0x1  │ .symtab         │
 │  ├──────┼─────────────────┤
 │  │ 0x9  │ .strtab         │
 │  ├──────┼─────────────────┤
 │  │ 0x11 │ .shstrtab       │
 │  ├──────┼─────────────────┤
 │  │ 0x1b │ .rela.text      │
 │  ├──────┼─────────────────┤
 │  │ 0x26 │ .data           │
 │  ├──────┼─────────────────┤
 │  │ 0x2c │ .bss            │
 │  ├──────┼─────────────────┤
 │  │ 0x31 │ .rodata         │
 │  ├──────┼─────────────────┤
 │  │ 0x39 │ .comment        │
 │  ├──────┼─────────────────┤
 │  │ 0x42 │ .note.GNU-stack │
 │  ├──────┼─────────────────┤
 │  │ 0x52 │ .rela.eh_frame  │
 │  ╰──────┴─────────────────╯
 ┴

section .text#2 (address: 0x0)
 │
 │  program data | permissions: R X
 │
 │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │  │ 55 48 89 e5 48 83 ec 10 89 7d fc 48 89 75 f0 bf │ UH..H....}.H.u.. │
 │  │ 00 00 00 00 e8 00 00 00 00 b8 00 00 00 00 c9 c3 │ ................ │
 │  ╰─────────────────────────────────────────────────┴──────────────────╯
 ┴

section .rela.text#3 (address: 0x0)
 │
 │  symbol table:       .symtab#4
 │  applies to section: .text#2
 │
 │    Relocations:
 │  ╭──────────────┬────────┬────────┬────────╮
 │  │ Type         │ Symbol │ Offset │ Addend │
 │  ├──────────────┼────────┼────────┼────────┤
 │  │ X86_64_32    │ #0     │ 0x10   │ 0x0    │
 │  ├──────────────┼────────┼────────┼────────┤
 │  │ X86_64_PLT32 │ puts#1 │ 0x15   │ -0x4   │
 │  ╰──────────────┴────────┴────────┴────────╯
 ┴

section .symtab#4 (address: 0x0)
 │
 │    Symbol table:
 │  ╭───────────┬─────────┬──────────┬────────────┬────────────┬───────┬──────╮
 │  │ Name      │ Binding │ Type     │ Visibility │ Definition │ Value │ Size │
 │  ├───────────┼─────────┼──────────┼────────────┼────────────┼───────┼──────┤
 │  │ #0        │ Local   │ Section  │ Default    │ .rodata#7  │ 0x0   │ 0x0  │
 │  ├───────────┼─────────┼──────────┼────────────┼────────────┼───────┼──────┤
 │  │ puts#1    │ Global  │ -        │ Default    │ Undefined  │ 0x0   │ 0x0  │
 │  ├───────────┼─────────┼──────────┼────────────┼────────────┼───────┼──────┤
 │  │ #2        │ Local   │ Section  │ Default    │ .text#2    │ 0x0   │ 0x0  │
 │  ├───────────┼─────────┼──────────┼────────────┼────────────┼───────┼──────┤
 │  │ #3        │ Local   │ -        │ Default    │ Undefined  │ 0x0   │ 0x0  │
 │  ├───────────┼─────────┼──────────┼────────────┼────────────┼───────┼──────┤
 │  │ hello.c#4 │ Local   │ File     │ Default    │ Absolute   │ 0x0   │ 0x0  │
 │  ├───────────┼─────────┼──────────┼────────────┼────────────┼───────┼──────┤
 │  │ main#5    │ Global  │ Function │ Default    │ .text#2    │ 0x0   │ 0x20 │
 │  ╰───────────┴─────────┴──────────┴────────────┴────────────┴───────┴──────╯
 ┴

section .data#5 (address: 0x0)
 │
 │  program data | permissions: RW
 │
 │  ╭──┬──╮
 │  │  │  │
 │  ╰──┴──╯
 ┴

section .bss#6 (address: 0x0)
 │
 │  uninitialized | len: 0x0 | permissions: RW 
 ┴

section .rodata#7 (address: 0x0)
 │
 │  program data | permissions: R
 │
 │  ╭─────────────────────────────────────┬──────────────╮
 │  │ 48 65 6c 6c 6f 20 77 6f 72 6c 64 00 │ Hello world. │
 │  ╰─────────────────────────────────────┴──────────────╯
 ┴

section .comment#8 (address: 0x0)
 │
 │  program data | permissions: -   | deduplicating zero-terminated strings
 │
 │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │  │ 00 47 43 43 3a 20 28 55 62 75 6e 74 75 20 31 31 │ .GCC: (Ubuntu 11 │
 │  │ 2e 34 2e 30 2d 31 75 62 75 6e 74 75 31 7e 32 32 │ .4.0-1ubuntu1~22 │
 │  │ 2e 30 34 29 20 31 31 2e 34 2e 30 00             │ .04) 11.4.0.     │
 │  ╰─────────────────────────────────────────────────┴──────────────────╯
 ┴

section .note.GNU-stack#9 (address: 0x0)
 │
 │  program data | permissions: -
 │
 │  ╭──┬──╮
 │  │  │  │
 │  ╰──┴──╯
 ┴

section .eh_frame#10 (address: 0x0)
 │
 │  program data | permissions: R
 │
 │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │  │ 14 00 00 00 00 00 00 00 01 7a 52 00 01 78 10 01 │ .........zR..x.. │
 │  │ 1b 0c 07 08 90 01 00 00 1c 00 00 00 1c 00 00 00 │ ................ │
 │  │ 00 00 00 00 20 00 00 00 00 41 0e 10 86 02 43 0d │ .... ....A....C. │
 │  │ 06 5b 0c 07 08 00 00 00                         │ .[......         │
 │  ╰─────────────────────────────────────────────────┴──────────────────╯
 ┴

section .rela.eh_frame#11 (address: 0x0)
 │
 │  symbol table:       .symtab#4
 │  applies to section: .eh_frame#10
 │
 │    Relocations:
 │  ╭─────────────┬────────┬────────┬────────╮
 │  │ Type        │ Symbol │ Offset │ Addend │
 │  ├─────────────┼────────┼────────┼────────┤
 │  │ X86_64_PC32 │ #2     │ 0x20   │ 0x0    │
 │  ╰─────────────┴────────┴────────┴────────╯
 ┴

section .strtab#12 (address: 0x0)
 │
 │    Strings table:
 │  ╭─────┬─────────╮
 │  │ 0x0 │         │
 │  ├─────┼─────────┤
 │  │ 0x1 │ hello.c │
 │  ├─────┼─────────┤
 │  │ 0x9 │ main    │
 │  ├─────┼─────────┤
 │  │ 0xe │ puts    │
 │  ╰─────┴─────────╯
 ┴

No segments in the ELF file.

no stderr present
